### PR TITLE
[DO-2033] Pdf pep

### PR DESCRIPTION
**Pull request by Iva-khar**

### CODE REVIEW

File: business_register/views/pep_views.py [hunk 1]
```diff
@@ -54,6 +57,12 @@ def retrieve_by_source_id(self, request, pk):
         serializer = self.get_serializer(pep)
         return Response(serializer.data)
 
+    @action(detail=True, url_path='pdf', serializer_class=PepDetailSerializer)
+    def pdf(self, request, pk):
+        pep = get_object_or_404(self.get_queryset(), source_id=pk)
+        serializer = self.get_serializer(pep)
+        return PDFViewMixin(serializer.data)
```
Comment (roman-tiukh) [inline business_register/views/pep_views.py:64]: redundant code

File: business_register/views/pep_views.py [hunk 2]
```diff
@@ -127,3 +136,15 @@ def check(self, request):
...
+class PepPDFView(PDFDetailView):
+    queryset = Pep.objects.all()
+    serializer_class = PepDetailSerializer
+    template_name = 'pdf/pep.html'
+
+    def get_template_names(self):
+        return self.template_name
+
+    def get_file_name(self, context):
+        return 'pep_' + str(context['id'])
```
Comment (roman-tiukh) [inline business_register/views/pep_views.py:150]:
```suggestion
        return f'pep_{self.object.id}.pdf'
```

File: business_register/views/pep_views.py [hunk 3]
```diff
@@ -127,3 +136,15 @@ def check(self, request):
...
         return Response(serializer.data)
+
+
+class PepPDFView(PDFDetailView):
+    queryset = Pep.objects.all()
+    serializer_class = PepDetailSerializer
+    template_name = 'pdf/pep.html'
+
+    def get_template_names(self):
+        return self.template_name
```
Comment (roman-tiukh) [inline business_register/views/pep_views.py:147]: delete this

File: business_register/views/pep_views.py [hunk 4]
```diff
@@ -127,3 +136,15 @@ def check(self, request):
         peps = self.filter_queryset(self.get_queryset())
         serializer = self.get_serializer(peps, many=True)
         return Response(serializer.data)
+
+
+class PepPDFView(PDFDetailView):
+    queryset = Pep.objects.all()
+    serializer_class = PepDetailSerializer
```
Comment (roman-tiukh) [inline business_register/views/pep_views.py:143]: delete this

File: business_register/views/pep_views.py [hunk 5]
```diff
@@ -127,3 +136,15 @@ def check(self, request):
...
+    template_name = 'pdf/pep.html'
```
Comment (roman-tiukh) [inline business_register/views/pep_views.py:144]:
```suggestion
    template_name = 'business_register/pep/pep_pdf.html'
```

File: business_register/views/pep_views.py [hunk 6]
```diff
@@ -20,7 +23,7 @@
 from data_converter.pagination import CachedCountPagination
 from data_ocean.permissions import IsAuthenticatedAndPaidSubscription
 from data_ocean.tasks import export_to_s3
-from data_ocean.views import CachedViewSetMixin, RegisterViewMixin
+from data_ocean.views import CachedViewSetMixin, RegisterViewMixin, PDFViewMixin, PDFDetailView
```
Comment (roman-tiukh) [inline business_register/views/pep_views.py:26]:
```suggestion
from data_ocean.views import CachedViewSetMixin, RegisterViewMixin, PDFDetailView
```

File: data_converter/urls.py [hunk 7]
```diff
@@ -131,6 +131,7 @@
 
     path('api/pep/scoring/<int:source_id>/', PepScoringDetailView.as_view()),
     path('api/', include((router.urls, 'registers'), namespace='registers')),
+    path('pdf/pep/<int:pk>/', PepPDFView.as_view())
```
Comment (roman-tiukh) [inline data_converter/urls.py:134]:
```suggestion
    path('api/pep/<int:pk>/pdf/', PepPDFView.as_view())
```

File: data_ocean/views.py [hunk 8]
```diff
@@ -381,3 +386,53 @@ def add_manual_parameters(self, parameters):
...
+        return FileResponse(file, content_type='application/pdf')
+
+
+class PDFDetailView(PDFViewMixin, RetrieveAPIView):
+    def get(self, request, *args, **kwargs):
+        self.object = self.get_object()
+        return super().get(request, *args, **kwargs)
+
+    def get_context_data(self, pk):
+        return model_to_dict(self.object)
```
Comment (roman-tiukh) [inline data_ocean/views.py:438]: Nooo, you don't need this

File: data_ocean/views.py [hunk 9]
```diff
@@ -381,3 +386,53 @@ def add_manual_parameters(self, parameters):
...
+        raise NotImplementedError
+
+    @staticmethod
+    def render_to_pdf(template_name, context: dict,
+                      file_name: str = None, changed_fonts: bool = False) -> io.BytesIO:
+        html = render_to_string(template_name, context)
+        file = io.BytesIO()
+        extra = {}
+        # if changed_fonts:
+        #     extra['font_config'] = FontConfiguration()
```
Comment (roman-tiukh) [inline data_ocean/views.py:404]: why you comment this???

File: business_register/views/pep_views.py [hunk 10]
```diff
@@ -127,3 +136,15 @@ def check(self, request):
         peps = self.filter_queryset(self.get_queryset())
         serializer = self.get_serializer(peps, many=True)
         return Response(serializer.data)
+
+
+class PepPDFView(PDFDetailView):
+    queryset = Pep.objects.all()
```
Comment (roman-tiukh) [inline business_register/views/pep_views.py:142]: Add permission_classes

**Review by roman-tiukh — CHANGES_REQUESTED** [review]

File: business_register/views/pep_views.py (hunk 1 above)
Comment (Iva-khar) [reply business_register/views/pep_views.py:64]: deleted

File: business_register/views/pep_views.py (hunk 2 above)
Comment (Iva-khar) [reply business_register/views/pep_views.py:150]: done

File: business_register/views/pep_views.py (hunk 3 above)
Comment (Iva-khar) [reply business_register/views/pep_views.py:147]: `AttributeError: 'PepPDFView' object has no attribute 'get_template_names'`

File: business_register/views/pep_views.py (hunk 6 above)
Comment (Iva-khar) [reply business_register/views/pep_views.py:26]: done

File: business_register/views/pep_views.py (hunk 4 above)
Comment (Iva-khar) [reply business_register/views/pep_views.py:137]: done

File: business_register/views/pep_views.py (hunk 5 above)
Comment (Iva-khar) [reply business_register/views/pep_views.py:138]: done

File: data_converter/urls.py (hunk 7 above)
Comment (Iva-khar) [reply data_converter/urls.py:134]: done

File: data_ocean/views.py (hunk 9 above)
Comment (Iva-khar) [reply data_ocean/views.py:404]: uncommented

File: business_register/views/pep_views.py (hunk 10 above)
Comment (Iva-khar) [reply business_register/views/pep_views.py:136]: done

File: data_ocean/views.py (hunk 8 above)
Comment (Iva-khar) [reply data_ocean/views.py]: rewrited

File: data_ocean/views.py [hunk 11]
```diff
@@ -381,3 +386,50 @@ def add_manual_parameters(self, parameters):
...
+        assert self.template_name, \
+            f'{self.__class__.__name__} dont have "template_name" attribute'
+        context = self.get_context_data(**kwargs)
+        file = self.render_to_pdf(
+            template_name=self.template_name,
+            context=context,
+            file_name=self.get_file_name(context),
+            changed_fonts=self.changed_fonts,
+        )
+        return FileResponse(file, content_type='application/pdf')
```
Comment (roman-tiukh) [inline data_ocean/views.py:429]: no get_context_data with user.

File: data_ocean/views.py [hunk 12]
```diff
@@ -381,3 +386,50 @@ def add_manual_parameters(self, parameters):
...
+
+
+class PDFDetailView(PDFViewMixin, RetrieveAPIView):
+    def get(self, request, *args, **kwargs):
+        self.object = self.get_object()
+        return super().get(request, *args, **kwargs)
```
Comment (roman-tiukh) [inline data_ocean/views.py:435]: no get_context_data with object

**Review by roman-tiukh — CHANGES_REQUESTED** [review]
no PermissionViewMixin.
Still in work

**Pull request closed without merging**